plugins/parcelport/verbs/readers_writers_mutex.hpp

```diff
@@ -0,0 +1,264 @@
+ *
+ * This has the effect of queuing lock requests in the order they arrive
+ * (incidentally avoiding starvation).
+ *
+ * Each lock/unlock pair requires incrementing both 'readers' and 'writers'.
+ * In the case of a reader, the 'readers' increment happens when the reader
+ * acquires the lock (to allow read-lock sharing), and the 'writers' increment
+ * happens when the reader releases the lock. In the case of a writer, both
+ * 'readers' and 'writers' are incremented when the writer releases the lock.
+ *
+ * For example, consider the following read (R) and write (W) lock requests:
+ *
+ *                                              writers readers next
+ *                                              0       0       0
+ *      R: ticket 0, readers match      OK      0       1       1
+ *      R: ticket 1, readers match      OK      0       2       2
+ *      R: ticket 2, readers match      OK      0       3       3
+ *      W: ticket 3, writers no match   block   0       3       4
+ *      R: ticket 2, unlock                     1       3       4
+ *      R: ticket 0, unlock                     2       3       4
+ *      R: ticket 1, unlock                     3       3       4
+ *      W: ticket 3, writers match      OK      3       3       4
+ *
+ * Note the writer blocks until 'writers' equals its ticket number and it does
+ * not matter if readers unlock in order or not.
+ *
+ * Readers or writers entering the system after the write lock is queued block,
+ * and the next ticket holder (reader or writer) will unblock when the writer
+ * unlocks. An example, continuing from the last line of the above example:
+ *
+ *                                              writers readers next
+ *      W: ticket 3, writers match      OK      3       3       4
+ *      R: ticket 4, readers no match   block   3       3       5
+ *      R: ticket 5, readers no match   block   3       3       6
+ *      W: ticket 6, writers no match   block   3       3       7
+ *      W: ticket 3, unlock                     4       4       7
+ *      R: ticket 4, readers match      OK      4       5       7
+ *      R: ticket 5, readers match      OK      4       6       7
+ *
+ * The 'next' field is a 2-byte value so the available ticket number wraps at
+ * 64K requests. If a thread's lock request is not granted until the 'next'
+ * field cycles and the same ticket is taken by another thread, we could grant
+ * a lock to two separate threads at the same time, and bad things happen: two
+ * writer threads or a reader thread and a writer thread would run in parallel,
+ * and lock waiters could be skipped if the unlocks race. This is unlikely, it
+ * only happens if a lock request is blocked by 64K other requests. The fix is
+ * to grow the lock structure fields, but the largest atomic instruction we have
+ * is 8 bytes, the structure has no room to grow.
+ */
+// from http://locklessinc.com/articles/locks/
+
```

Comment:
My plan was to migtrate this to lcos and the concurrent map to hpx::concurrent, but only after this is merged to master and work on hpx::concurrent is more mature.